paper-ripple.html

```diff
@@ -116,362 +113,418 @@
-      return Math.max(0, Math.min(outerOpacity, waveOpacity));
-    }
+      distance: function (x1, y1, x2, y2) {
+        var xDelta = (x1 - x2);
+        var yDelta = (y1 - y2);
 
-    // Determines whether the wave should be completely removed.
-    function waveDidFinish(wave, radius, anim) {
-      var waveOpacity = waveOpacityFn(wave.tDown, wave.tUp, anim);
+        return Math.sqrt(xDelta * xDelta + yDelta * yDelta);
+      },
+
+      now: (function () {
+        if (window.performance && window.performance.now) {
+          return window.performance.now.bind(window.performance);
+        }
 
-      // If the wave opacity is 0 and the radius exceeds the bounds
-      // of the element, then this is finished.
-      return waveOpacity < 0.01 && radius >= Math.min(wave.maxRadius, waveMaxRadius);
+        return Date.now;
+      })()
     };
 
-    function waveAtMaximum(wave, radius, anim) {
-      var waveOpacity = waveOpacityFn(wave.tDown, wave.tUp, anim);
+    function ContainerMetrics (element) {
+      this.element = element;
+      this.width = this.boundingRect.width;
+      this.height = this.boundingRect.height;
 
-      return waveOpacity >= anim.initialOpacity && radius >= Math.min(wave.maxRadius, waveMaxRadius);
+      this.size = Math.max(this.width, this.height);
     }
 
-    //
-    // DRAWING
-    //
-    function drawRipple(ctx, x, y, radius, innerAlpha, outerAlpha) {
-      // Only animate opacity and transform
-      if (outerAlpha !== undefined) {
-        ctx.bg.style.opacity = outerAlpha;
+    ContainerMetrics.prototype = {
+      get boundingRect () {
+        return this.element.getBoundingClientRect();
+      },
+
+      furthestCornerDistanceFrom: function (x, y) {
+        var topLeft = Utility.distance(x, y, 0, 0);
+        var topRight = Utility.distance(x, y, this.width, 0);
+        var bottomLeft = Utility.distance(x, y, 0, this.height);
+        var bottomRight = Utility.distance(x, y, this.width, this.height);
+
+        return Math.max(topLeft, topRight, bottomLeft, bottomRight);
       }
-      ctx.wave.style.opacity = innerAlpha;
+    };
 
-      var s = radius / (ctx.containerSize / 2);
-      var dx = x - (ctx.containerWidth / 2);
-      var dy = y - (ctx.containerHeight / 2);
+    function Ripple (element) {
+      this.element = element;
+      this.color = window.getComputedStyle(element).color;
 
-      ctx.wc.style.webkitTransform = 'translate3d(' + dx + 'px,' + dy + 'px,0)';
-      ctx.wc.style.transform = 'translate3d(' + dx + 'px,' + dy + 'px,0)';
+      this.wave = document.createElement('div');
+      this.waveContainer = document.createElement('div');
+      this.wave.style.backgroundColor = this.color;
+      this.wave.classList.add('wave');
+      this.waveContainer.classList.add('wave-container');
+      this.waveContainer.appendChild(this.wave);
+      this.container.appendChild(this.waveContainer);
+      this.background.style.backgroundColor = this.color;
```

Comment:
I agree, it's not totally obvious what that refers to. On the other hand, in my opinion it's incredibly unsanitary for this class (which should be an element in its own right) to be reaching into the element this way in the first place. My intention here is to simplify the existing implementation, so that an obvious, sanitary way forward can emerge.